### ISSUE TITLE
Generate sys module

### ISSUE THREAD

**Issue by EPashkin** [gkoz/gir#29]
It you don't start it, i can begin do it (about  1 day) in folowing way:
Add in Gir.toml `options.sys_subpath = "gtk-sys"`
Generate funcs.rs with `status = "generate"` widgets functions, that will be included and reexported in lib.rs.
All sys-depended code will go to `sys` subdirs of `analyze` and `codegen`

**Comment by EPashkin** [comment]
Also "return out parameters" little depends on this.

**Comment by gkoz** [comment]
I do plan to start the sys crate generation but why should anything else depend on that?

**Comment by EPashkin** [comment]
You still planing, or already started?
In "return out parameters" I planing add `rets:Vec<Parameters>` along with old `ret: Option<Parameter>` to `object::Info`.
I want this issue act as warning remover, therefore, I propose do it myself

**Comment by gkoz** [comment]
No, I have not started that. But I didn't want to couple the sys crate generation with other modes too much, for instance, I don't see why it should care about which widgets and methods are generated.

**Comment by EPashkin** [comment]
If you want generate all function, then no dependence.

**Comment by EPashkin** [comment]
gkoz: if you don't begin generate sys module, can i generate functions for all classes?

**Comment by gkoz** [comment]
Alright, go ahead.

**Comment by EPashkin** [comment]
gkoz: can you add to gtk-sys file funcs.rs where move `GtkAboutDialog` functions?
I have problem with this.

**Comment by gkoz** [comment]
I don't see the point of splitting the sys crates into modules especially if they're autogenerated.

**Comment by EPashkin** [comment]
Then where i write functions?

**Comment by EPashkin** [comment]
Or you want autogenerate all sys module from begining?

**Comment by gkoz** [comment]
Yes, I could be wrong but this task seems simple enough to generate everything at once.

**Comment by EPashkin** [comment]
Ok. Then in first version i just rewrite begining of lib.rs.

**Comment by gkoz** [comment]
:+1:

**Comment by gkoz** [comment]
I think we're almost feature-complete vs the handmade sys crates so it's time to eliminate the gtk-related hardcoded stuff and accomodate generating all of `glib-sys`, `gdk-sys`, (possbily `pango-sys`), and `gtk-sys`.

**Comment by EPashkin** [comment]
Any hardcode in sys?
IMHO better start with `glib-sys`. we need generate `lib.rs, glib.rs, gobject.rs, gio.rs`?

**Comment by EPashkin** [comment]
Not sure what better, one `Gir.toml` for each project, that can be stored in project,
or one config for all libs, that stored in `gir`

**Comment by gkoz** [comment]
One way could be to generate `glib.rs`, `gio.rs` etc and do glob reexports in the `lib.rs` (`pub use glib::*;`). We'd also need a manual module with `gboolean`, `gsize` and other typedefs that could be imported uniformly from every generated module.

**Comment by EPashkin** [comment]
Ok, I try make template tomorrow

**Comment by EPashkin** [comment]
What harcoded in sys?

**Comment by EPashkin** [comment]
IMHO only file path, and `fix_namespace`, it will changed anyway

**Comment by gkoz** [comment]
A part of the `statics.rs` too. Maybe that's all that is hardcoded.

**Comment by gkoz** [comment]
As to the configs, currently I prefer to have a separate config for each lib (namespace), for this choosing a config from command line would be handy. The configs might need to support blacklisting types and definitely need blacklisting constants because `GLib-2.0.gir` has a bunch of platform-specific ones.

**Comment by gkoz** [comment]
Let's just save ourselves the trouble and make `glib-sys`, `gio-sys` etc separate crates. This will have very little impact on the "front end" crates anyway.

**Comment by EPashkin** [comment]
If @GuillaumeGomez allow it

**Comment by GuillaumeGomez** [comment]
As long as it doesn't break rust-gnome "environment", there's no problem for me.

**Comment by GuillaumeGomez** [comment]
Actually, if someone could explain me a bit longer what's happening here, it would be amazing ! If you want to generate crates to be bound by all rust-gnome libraries, shouldn't we run tests before ? Otherwise, I can give write access to crates to @gkoz.

**Comment by gkoz** [comment]
@GuillaumeGomez
The crates generated by this utility aren't yet ready to replace the rust-gnome ones (actually, only the `glib-sys` crate builds cleanly so far) but they're getting there and I plan to closely review the differences when that time comes. Testing the stuff properly is a tall order though because the coverage our `examples` crate provides is rather modest :)

The task will be less cumbersome if there is a 1:1 match between the libraries (such as `glib`, `gio`, `gobject`) and crates, so the generator won't have to incorporate another bunch of special cases. This doesn't make it necessary to change the higher level crates organization. But we might consider putting the new sys crates into their own single `sys` (or `ffi`) repo and removing the sys crates from the rest of the repos.

**Comment by GuillaumeGomez** [comment]
We actually could already do it. I think it's a bit disturbing for newcomers. We didn't have a big enough use to split `sys` from its "parent". Maybe it's the time ?

**Comment by gkoz** [comment]
In a week we might have a prototype repo with newly generated sys crates for everything except cairo.

**Comment by GuillaumeGomez** [comment]
Perfect, I'll wait until this date then. :-)

**Comment by EPashkin** [comment]
For full replace current syses we also need `feature` generation and better with deprecation.

**Comment by gkoz** [comment]
The current sys crates don't really do anything with the features, if I remember right. But we might make the build script add some constant that will help downstream crates change their cfgs.

**Comment by EPashkin** [comment]
When i tried apply replace sys in gtk it errored about features https://github.com/rust-gnome/gtk/blob/master/Cargo.toml

**Comment by gkoz** [comment]
The features are present in gtk-sys' `Cargo.toml` but don't do anything. Since it just works, we can instead stop expecting the sys crates to have them.

**Comment by EPashkin** [comment]
IMHO generate build script unneeded functionality, as it almost never changed, so not need to update from *.gir. Also in manual way can be checked needed library in os specified way (ex. for `Gio-2.0.gir` it can always check `gio-2.0` and only on linux check `gio-unix-2.0`).

**Comment by gkoz** [comment]
I've finally found all the pieces need to replace the `feature`s with `cfg`s set by the build scripts, will make a proof of concept soon. The build script will be responsible for detecting the available library version and setting the `cfg`s that describe it (similar to current features).

**Comment by EPashkin** [comment]
Good.
You will create issue for discussing "features" in https://github.com/rust-gnome/gtk ?

**Comment by gkoz** [comment]
Yes.

**Comment by EPashkin** [comment]
Good work on autodetecting features.
Yet still I do not see the point to auto-generate build script.

**Comment by gkoz** [comment]
For one, they turned out big enough that not using some kind of template is going to be problematic.

**Comment by gkoz** [comment]
Potentially even `Cargo.toml` could be partly generated. Since toml is structured, it's possible to rewrite select portions while preserving manual modifications.

**Comment by EPashkin** [comment]
Agree, but IMHO as with build scripts it better done later, after we make main functionality.
We still don't have signals.

**Comment by EPashkin** [comment]
Found attribute `introspectable` in `GLib-2.0.gir`. Don't we need ignore this function?
```
      <function name="append_vals"
                c:identifier="g_array_append_vals"
                introspectable="0">
```

**Comment by gkoz** [comment]
Yes, the auxiliary files can be left for later. I guess the build scripts could be rewritten to be more easily copyable by hand. But they're not very hard to generate either. (BTW does the generator still fail to create the target directory e.g. `pango-sys`?)

`introspectable` has to do with dynamic introspection used by dynamic languages that (I believe) process at runtime the type libraries, which are binary platform-specific versions of the GIR. We can ignore it.

**Comment by EPashkin** [comment]
Creating dir fixed with removing it from test.sh

**Comment by EPashkin** [comment]
Seems I totally misunderstand what we doing with "features".
We add '#[cfg(gtk_3_10)]` to sys crates code or this condition added only to "normal" crates?

**Comment by gkoz** [comment]
Ultimately both. It's essential in normal crates and nice to have the sys ones.

**Comment by EPashkin** [comment]
Marked as done build script and features

**Comment by EPashkin** [comment]
"foreign" modules still need solution?

**Comment by gkoz** [comment]
Yes, I'd like to tackle `cairo` eventually. But the higher priority would be creating a united sys repo and migrating the official crates to depend on it. I plan start looking into this after adding some `Cargo.toml` conveniences. 

Speaking of `Cargo.toml`, the only annoyance I'm not sure how to solve cleanly yet is avoiding the need to manage cargo package versions manually (e.g. generally, `gobject-sys` needs to know the latest `glib-sys` version to put in the `dependency` section).

**Comment by gkoz** [comment]
So, check out https://github.com/gkoz/rust-gnome-sys
It includes this repo and `gir-files` as submodules and contains the generated crates. Call `make` to regenerate.
@EPashkin @GuillaumeGomez

**Comment by EPashkin** [comment]
Better add `.gitignore` for *.bak.

**Comment by gkoz** [comment]
Are the backup files helping us anywhere? Don't we have superior version control through git?

**Comment by EPashkin** [comment]
For syses it really unneeded.

**Comment by gkoz** [comment]
I've pushed the (mostly mechanical) changes to [glib](https://github.com/gkoz/rust-gnome-glib/tree/new_sys), [pango](https://github.com/gkoz/rust-gnome-pango/tree/new_sys) and [gdk](https://github.com/gkoz/rust-gnome-gdk/tree/new_sys) that adapt them to the automatically generated sys crates.
As usual, everything at once can be checked out with
```shell
> repo init -u https://github.com/gkoz/rust-gnome-manifest -b new_sys
```

**Comment by GuillaumeGomez** [comment]
Wo, big change and work ! Can't wait to see this merged ! :-)

**Comment by gkoz** [comment]
I forgot about this kludge. It's blocking the windows build
https://github.com/rust-gnome/gdk/commit/cfc213e281d82ed322e0cdec5f77d6a00d28d08f

**Comment by gkoz** [comment]
This could surface somewhere else so we might add a `windows_utf8_shims` array to the config that would list the functions that need that treatment. Or just hardcode it.

**Comment by gkoz** [comment]
Hmm, actually, if we ever started processing GIRs generated on Windows, I'm not sure how these defines would be treated:
https://github.com/GNOME/gdk-pixbuf/blob/master/gdk-pixbuf/gdk-pixbuf-core.h#L269-L271
It could be that the the sys crate is supposed to define `gdk_pixbuf_new_from_file_utf8` on Windows and then the consumer crate must use `cfg(windows)` to choose the right ffi function.

**Comment by GuillaumeGomez** [comment]
I'll start to help here in the next week. Thanks for you awesome work @gkoz and @EPashkin !

**Comment by gkoz** [comment]
I've added some bullet points.

The struct sizes issue appears more important now because we need to make `sizeof(GValue)` correct (our Rust code wraps it by value) and it currently isn't on 32-bit platforms.
